### ISSUE TITLE
spritec-preview is broke

### ISSUE THREAD

**Issue by daose** [ProtoArt/spritec#73]
reproduce: run `cargo run --package spritec-preview` on master

I know we are scrapping this and replacing it with a proper GUI, but just wanted to point out that #58 might have broke spritec-preview in case anyone was relying on it

**Comment by sunjay** [comment]
Thanks for reporting this! Given that CI didn't catch it I'm guessing that it builds but doesn't run? What's the error message?

**Comment by daose** [comment]
whoops should've included this. No error messages, it just look like this:
![broke](https://user-images.githubusercontent.com/15278132/63732439-f01bf080-c8a6-11e9-99c9-f82fc5a18d8b.gif)

I'm guessing it's something to do with camera angle?

**Comment by sunjay** [comment]
This is pretty wild. I don't see anything in that PR that would create anything like this. Could you try checking out https://github.com/ProtoArt/spritec/commit/f85ff5819165c181e55e7fe8e52177ae3f167a8d to see if you can reproduce this?

Also, @JLSJamesShi could you check out master and try to run the command from the top comment in this issue? Do you get the same thing?

**Comment by daose** [comment]
f85ff58 looks fine like normal

**Comment by sunjay** [comment]
Hmm okay. I think if you make the changes to spritec-preview in https://github.com/ProtoArt/spritec/commit/acc837a302cd888dcda51b6ccd23397a4d739af5 it will fix the issues. (I'm guessing the camera is inside the model right now.) Please don't commit anything to that file though because it'll cause merge conflicts for my UI work. We'll close this issue once the current UI code is replaced. (It's good to have something around to document this workaround.)

**Comment by marauder00** [comment]
I can take a closer look at this if you want...my guess is because clipping isn't implemented in our software rasterizer

**Comment by daose** [comment]
up to you, I'm just using the CLI right now to see the preview so not relying on this

**Comment by sunjay** [comment]
Now that I've had some time to step away from this for a while, I'm far less attached to the way it is. I'm totally fine with switching completely to OpenGL as long as we are willing to accept the liabilities that come with that. At least Mara and I will need to take on the learning curve of OpenGL and making sure we remain compatible with WebGL in order to keep our options open. We should only switch if we're willing to take on that burden. 

This isn't super relevant to this issue, but the software rendering library came up so I thought I'd mention it. 

P.S. It may be worth trying out euc 0.4 to see if it addresses any of our issues. They switched to a proper clipping implementation and you can now return 4D vectors from the vertex shader.

**Comment by marauder00** [comment]
I don't want to end up rebuilding the OpenGL pipeline to get all of our necessary rendering features working :D Especially when working with materials, textures, and post-processing effects, OpenGL's features will make the development of these features easier. I think these pros, including platform popularity and developer support, outweigh the initial setup cost. 

TLDR, +1 on switching to OpenGL. I will checkout out euc 0.4 nevertheless.

**Comment by sunjay** [comment]
Sounds good! I'm ready to pull the trigger on whatever we end up deciding. I just want our progress to speed up. :D

**Comment by marauder00** [comment]
Thank you so much for looking into this btw ^_^!!!

**Comment by sunjay** [comment]
@marauder00 The best way for us to switch to OpenGL in my opinion is to start to use [glium](https://github.com/glium/glium). That crate provides a really nice wrapper around OpenGL that avoids a lot of the unsafety and common issues that people encounter while still giving you the full power of OpenGL. You can ignore the notice at the top that makes it seem like it's unmaintained. It's actually still very actively maintained, just not by the main person who started the crate. I don't even know why they keep that notice there frankly.

Do you want to start the work to port us to OpenGL? I can open an issue for it and provide some kind of roadmap if that would be helpful. (I haven't looked into it much, but I can do that work if you think it would help you.)

**Comment by sunjay** [comment]
spritec-preview is now deleted! :tada:
